File: leetcode/198-house-robber/house-robber.cpp
class Solution {
public:
    int __rob(vector<int> &nums,int index,vector<int> &memo){
        if(index == 0){
            return nums[0];
        }
        if(index < 0){
            return 0;
        }
        if(memo[index] != -1){
            return memo[index];

        }
        int pick = nums[index] + __rob(nums,index-2,memo);
        int notpick = __rob(nums,index-1,memo);
        memo[index] = max(pick,notpick);
        return memo[index];
    }
    int rob(vector<int>& nums) {
        int len=nums.size();
        vector<int> memo(len+1,-1);
        return __rob(nums,len-1,memo);
    }
};

File: leetcode/198-house-robber/house-robber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        i = 0
        dp = [-1]*(len(nums)+1)
        dp[0] = nums[0]
        
        n = len(nums)
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[1],nums[0])
        dp[1] = max(nums[1],nums[0])
        for i in range(2,len(nums)):
            dp[i] = max(nums[i] + dp[i-2],dp[i-1])
        
        return dp[n-1]
